Add CONTRIBUTING.md (#19257)
Azure/azure-rest-api-specs: CONTRIBUTING.md

## Design Guidelines

Azure REST APIs should conform to the [Azure REST API Guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md).

There is a [YouTube video series](https://www.youtube.com/watch?v=9Ng00IlBCtw) by Jeffrey Richter that describes the guidelines in detail.

Another resource is the [Considerations for Service Design](https://github.com/microsoft/api-guidelines/blob/vNext/azure/ConsiderationsForServiceDesign.md), which is an introduction to REST API design mainly for services that are just getting started.

### Exceptions for consistency within a service

There are situations where a service has GA'd their API with design patterns that do not follow our guidelines and it would be a breaking change to adopt the API design in the guidelines.
Because the first rule is to avoid breaking changes and because we want APIs to be consistent within a service, these design patterns are considered the standard for that service and should be followed in all subsequent (non-breaking) versions of that service's REST API.

## Coding Style

> REST APIs for Azure should be defined using the [OpenAPI v2](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md) format.

Reviewer: wonder if we should recommend Cadl here for Cadl libraries contain rules that encourage following established patterns for Azure APIs.  The document can be referred is https://github.com/Azure/cadl-azure/blob/main/docs/cadl-getting-started.md

## Pull Requests

If you want to contribute to the repository, follow these steps:  
  1. Fork the repository and create a new branch for your changes.
  2. If you are introducing a new api-version, create a new directory for that api-version and copy all the files from the previous version into the new directory. Make this the very first commit in your branch and then make your changes in subsequent commits.
  3. Use the [linting tools](#coding-style) to check your changes for compliance with the OpenAPI v2 standard, the Azure REST API Guidelines and that Azure API Style Guide.
  4. Push the changes to the branch in your fork until the branch is ready to be integrated.
  5. Rebase your branch if needed to incorporate any changes to **main** and submit a pull request to the **main** branch using either the "control plane" or "data plane" PR template -- the main PR template lets you select which one.
  5. Resolve any issues flagged by the [Pull Request checks](#pull-request-checks).
  6. Contact the ARM API Review board or Azure API Stewardship team to request PR review. 

## Pull Request checks

> Every PR in this repo will go through a series of PR checks, including:

Reviewer: Or https://aka.ms/ci-fix can be referred for how to enable API spec validation locally.

> Every PR in this repo will go through a series of PR checks, including:

Author: Added.

# Contributing

This file provides general guidance for developers that are creating or updating REST API definitions for Azure.

## Table of Contents
<!--
  You should regenerate the TOC using the `markdown-toc` node package after making changes to this file.

      npx markdown-toc -i --maxdepth 4 CONTRIBUTING.md
  -->

> <!-- toc -->

Reviewer 2: Just a reminder to regen toc before merge.

Author: Thanks for the reminder!  Done.

- [Design Guidelines](#design-guidelines)
- [Coding Style](#coding-style)
- [Directory Structure](#directory-structure)
- [Pull Requests](#pull-requests)
- [Adding a new service](#adding-a-new-service)
- [Running tests](#running-tests)
  * [Unit tests](#unit-tests)
  * [Integration tests](#integration-tests)
- [Developer's Certificate of Origin 1.1](#developers-certificate-of-origin-11)
- [Additional Resources](#additional-resources)

<!-- tocstop -->

## Reporting problems

If you discover a problem with a REST API document in this repo, feel free to [open an issue](https://github.com/Azure/azure-rest-api-specs/issues/new). But please do not report issues with service behavior / functionality in this repo.

When opening an issue, please clearly describe the problem including which REST API definition is in error and what elements of the definition are incorrect or missing.

## Avoid Breaking Changes

> The first rule for Azure REST APIs is to avoid breaking changes in Generally Available (GA'd) APIs. The [Azure Breaking Changes Policy](aka.ms/AzBreakingChangesPolicy) is a formal description of what changes are considered breaking changes. This rule trumps adherence to Design Guidelines, Coding Style, etc.

Reviewer 2: Use `https://aka.ms/AzBreakingChangesPolicy`, but it appears to be internal doc.

```suggestion
The first rule for Azure REST APIs is to avoid breaking changes in Generally Available (GA'd) APIs. The [Azure Breaking Changes Policy](https://aka.ms/AzBreakingChangesPolicy) is a formal description of what changes are considered breaking changes. This rule trumps adherence to Design Guidelines, Coding Style, etc.
```

### Exceptions for consistency within a service

There are situations where a service has GA'd their API with design patterns that do not follow our guidelines and it would be a breaking change to adopt the API design in the guidelines.
Because the first rule is to avoid breaking changes and because we want APIs to be consistent within a service, these design patterns are considered the standard for that service and should be followed in all subsequent (non-breaking) versions of that service's REST API.

## Coding Style

REST APIs for Azure should be defined using the [OpenAPI v2](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md) format.

The REST API should also adhere to the conventions documented in the [Azure API Style Guide](https://github.com/Azure/azure-api-style-guide/blob/main/openapi-style-guide.md).

You can use the [Spectral linter](https://meta.stoplight.io/docs/spectral/674b27b261c3c-overview) with our [Spectral ruleset](https://github.com/Azure/azure-api-style-guide#how-to-use-the-spectral-ruleset) to check conformance with OpenAPI v2 and the Style Guide.

> 

Reviewer 2: Should we mention about `npm prettier`?

## Directory Structure

> See the [README.md][./README.md] for a description of the directory structure to use when adding files to this repo.

Reviewer 2: link on `[README.md][./README.md]` not work. Should it be `[README.md](./README.md)`?

> See the [README.md][./README.md] for a description of the directory structure to use when adding files to this repo.

Author: ```suggestion
See the [README.md](./README.md) for a description of the directory structure to use when adding files to this repo.
```

## Pull Requests

If you want to contribute to the repository, follow these steps:  
  1. Fork the repository and create a new branch for your changes.
  2. If you are introducing a new api-version, create a new directory for that api-version and copy all the files from the previous version into the new directory. Make this the very first commit in your branch and then make your changes in subsequent commits.
  3. Use the [linting tools](#coding-style) to check your changes for compliance with the OpenAPI v2 standard, the Azure REST API Guidelines and that Azure API Style Guide.

>   4. Push to changes to the branch in your fork until the changes are ready to be integrated.

Reviewer 2: ```suggestion
  4. Push the changes to the branch in your fork until the branch is ready to be integrated.
```

  3. Use the [linting tools](#coding-style) to check your changes for compliance with the OpenAPI v2 standard, the Azure REST API Guidelines and that Azure API Style Guide.
  4. Push the changes to the branch in your fork until the branch is ready to be integrated.
  5. Rebase your branch if needed to incorporate any changes to **main** and submit a pull request to the **main** branch using either the "control plane" or "data plane" PR template -- the main PR template lets you select which one.
  5. Resolve any issues flagged by the [Pull Request checks](#pull-request-checks).
  6. Contact the ARM API Review board or Azure API Stewardship team to request PR review. 

Microsoft employees can try out the experience at [OpenAPI Hub](https://aka.ms/openapihub) for [adding a new API version using OpenAPI Hub](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

## Pull Request checks

Every PR in this repo will go through a series of PR checks, including:

- Breaking Changes - checks for incompatible changes in an existing api-version

> - Cross Version Breaking Changes - checks for changes in a new api-version that are incompatible with prior versions

Author: ```suggestion
- Breaking Changes (Cross-Version) - checks for changes in a new api-version that are incompatible with prior versions
```

# Contributing

This file provides general guidance for developers that are creating or updating REST API definitions for Azure.

## Table of Contents
<!--
  The TOC below is generated using the `markdown-toc` node package.

      https://github.com/jonschlinkert/markdown-toc

  You should regenerate the TOC after making changes to this file.

>       markdown-toc -i --maxdepth 4 CONTRIBUTING.md

Reviewer 3: You could just abbreviate this to have users run:
```bash
npx markdown-toc -i --maxdepth 4 CONTRIBUTING.md
```
Anyone that knows `npx` will now that you can run `npm info markdown-toc` to see info. `npx` takes care of pulling the package and effectively running `npm run` on it.

Just a thought.

>   The TOC below is generated using the `markdown-toc` node package.
> 
>       https://github.com/jonschlinkert/markdown-toc
> 
>   You should regenerate the TOC after making changes to this file.
> 
>       markdown-toc -i --maxdepth 4 CONTRIBUTING.md

Author: ```suggestion
  You should regenerate the TOC using the `markdown-toc` node package after making changes to this file.

      npx markdown-toc -i --maxdepth 4 CONTRIBUTING.md
```

  -->

<!-- toc -->

- [Design Guidelines](#design-guidelines)
- [Coding Style](#coding-style)
- [Directory Structure](#directory-structure)
- [Pull Requests](#pull-requests)
- [Adding a new service](#adding-a-new-service)
- [Running tests](#running-tests)
  * [Unit tests](#unit-tests)
  * [Integration tests](#integration-tests)
- [Developer's Certificate of Origin 1.1](#developers-certificate-of-origin-11)
- [Additional Resources](#additional-resources)

<!-- tocstop -->

## Reporting problems

If you discover a problem with a REST API document in this repo, feel free to [open an issue](https://github.com/Azure/azure-rest-api-specs/issues/new). But please do not report issues with service behavior / functionality in this repo.

When opening an issue, please clearly describe the problem including which REST API definition is in error and what elements of the definition are incorrect or missing.

## Avoid Breaking Changes

> The first rule for Azure REST APIs is to avoid breaking changes in GA'd APIs. The [Azure Breaking Changes Policy](aka.ms/AzBreakingChangesPolicy) is a formal description of what changes are considered breaking changes. This rule trumps adherence to Design Guidelines, Coding Style, etc.

Reviewer 3: I'd define acronyms the first time - at least per H2.
```suggestion
The first rule for Azure REST APIs is to avoid breaking changes in Generally Available (GA'd) APIs. The [Azure Breaking Changes Policy](aka.ms/AzBreakingChangesPolicy) is a formal description of what changes are considered breaking changes. This rule trumps adherence to Design Guidelines, Coding Style, etc.
```

Because the first rule is to avoid breaking changes and because we want APIs to be consistent within a service, these design patterns are considered the standard for that service and should be followed in all subsequent (non-breaking) versions of that service's REST API.

## Coding Style

REST APIs for Azure should be defined using the [OpenAPI v2](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md) format.

The REST API should also adhere to the conventions documented in the [Azure API Style Guide](https://github.com/Azure/azure-api-style-guide/blob/main/openapi-style-guide.md).

You can use the [Spectral linter](https://meta.stoplight.io/docs/spectral/674b27b261c3c-overview) with our [Spectral ruleset](https://github.com/Azure/azure-api-style-guide#how-to-use-the-spectral-ruleset) to check conformance with OpenAPI v2 and the Style Guide.

## Directory Structure

See the [README.md][./README.md] for a description of the directory structure to use when adding files to this repo.

## Pull Requests

If you want to contribute to the repository, follow these steps:  
  1. Fork the repository and create a new branch for your changes.

>   2. If you are introducing a new api-version, create a new directory for that api-version and copy all the files from the previous version into the new directory. Make this the very first commit in your branch and then make your changes in subsequent commits.

Reviewer 3: OpenAPI Hub can do this. Should we have them use that? It also updates the api-version for them. That said, the automatic stuff it does aren't always well-understood if people haven't used them.

Reviewer: OpenAPI Hub can do this, but the tools require MS alias to login, and the introduction wiki is in Azure SDK internal wiki repo, which is only accessible by MS employee.   Proposed wording could be 

Microsoft employees can try out the experience at [OpenAPI Hub](https://aka.ms/openapihub) for [adding a new API version using OpenAPI Hub](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

Author: Added.

  3. Use the [linting tools](#coding-style) to check your changes for compliance with the OpenAPI v2 standard, the Azure REST API Guidelines and that Azure API Style Guide.

>   4. Push to your fork and submit a pull request to the **main** branch using either the "control plane" or "data plane" PR template -- the main PR template lets you select which one.

Reviewer 3: This should only be when done. I question how common this is a one-and-done process. I highly recommend people - as I did in our internal docs - to create a feature branch instead e.g., `release-{service}-{RP}-{api-version(-preview)}` they target until they are all done, then submit a PR to main. However, because these feature branches tend to live for a while, they often need to merge main which gets *extremely* messy - hundreds of commits and thousands of unrelated files. Instead, they should rebase, which my instructions cover.

Lots of teams target main like this, and it's wrong because - whether in a preview/ or stable/ directory - they are likely to make breaking changes thinking that, "We haven't released it yet, so it's okay to change what's in main." According to past conversations with @JeffreyRichter and @johanste this is incorrect. Being in main means published - effectively immutable.

Author: ```suggestion
  4. Push to changes to the branch in your fork until the changes are ready to be integrated.
  5. Rebase your branch if needed to incorporate any changes to **main** and submit a pull request to the **main** branch using either the "control plane" or "data plane" PR template -- the main PR template lets you select which one.
```
